server/mqtt.go

```diff
@@ -3263,7 +3263,15 @@ func (s *Server) mqttCheckPubRetainedPerms() {
 		perms := map[string]*perm{}
 		deletes := map[string]uint64{}
 		asm.mu.Lock()
-		for subject, rm := range asm.retmsgs {
+		for subject, rf := range asm.retmsgs {
+			jsm, err := asm.jsa.loadMsg(mqttRetainedMsgsStreamName, rf.sseq)
```

Comment:
I am a bit concerned that we are going to load messages (possibly lot of them) under the sm and asm locks. Also, since retmsgs is a map, iteration may be causing lots of read out-of-order (sequence wise) and cause blocks to be loaded excessively.
So I propose that we do the process in several passes. Maybe collect the "asm" (sessions) under sm's lock. Then, iterate over those collected sessions and under asm's lock, collect the sequences in an array, release lock, then do an ascending sort of those sequences. For each sequence, do what you have here. For the asm.sl.Remove() down there, maybe capture the asm's sl under the lock (when collecting sequences).

If this is not clear at all, let me know and I can try to code it up.

Reply:
Good point on the locks. I can review it when I'm back on Tuesday!

Reply:
Done — if you don't mind, please check this is what you had in mind.

> For each sequence, do what you have here. For the asm.sl.Remove() down there, maybe capture the asm's sl under the lock (when collecting sequences).

`Remove` takes the lock itself in this case, so there's no need for me to do so.